### PR TITLE
refactor: move and document getNpmAuthenticationHeader

### PR DESCRIPTION
**Pull request by paul-soporan**
**What's the problem this PR addresses?**

The `getNpmAuthenticationHeader` hook was defined inside a `Hooks.ts` file instead of the index. Because of this, the hook extraction script didn't see that it wasn't documented.

**How did you fix it?**

Moved it to `index.ts` and added more checks to make sure that this can never happen again.

**Checklist**

- [X] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [X] I have set the packages that need to be released for my changes to be effective.

- [X] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: scripts/extract-hooks.ts [hunk 1]
```diff
@@ -110,6 +121,21 @@ async function execute(files: Array<string>) {
...
+  });
+
+  const processPromises = [];
+  for (const absolutePath of allFiles)
+    processPromises.push(Promise.resolve().then(() => parseFile(absolutePath).then(file => processFile(file, {allowHooks: false}))));
+
+  await Promise.all(processPromises);
+}
+
+exports.checkForInvalidHookLocations = checkForInvalidHookLocations;
```
Comment (arcanis) [inline scripts/extract-hooks.ts:137]: While it was indeed an issue, I'm not sure adding a runtime check on all TS/TSX files is a good idea. If really needed, I feel like adding a lint rule preventing the use of `Hooks` as interface name (possibly via `@typescript-eslint/naming-convention`, which seems to allow regexps) would be enough.

File: .eslintrc.js [hunk 2]
```diff
@@ -3,7 +3,25 @@ module.exports = {
     require.resolve(`@yarnpkg/eslint-config`),
     require.resolve(`@yarnpkg/eslint-config/react`),
   ],
-  ignorePatterns: [
-    `packages/plugin-compat/extra/fsevents/fsevents-*.js`,
+  overrides: [
+    {
+      files: [`!packages/*/sources/{index,Plugin}.ts`],
```
Comment (paul-soporan) [inline .eslintrc.js:8]: While the correct value would be `['!packages/*/sources/index.ts', '!packages/yarnpkg-core/sources/Plugin.ts']`, ESLint seems to still match the files if I put both negated patterns, even though [the docs](https://eslint.org/docs/user-guide/configuring/configuration-files#how-do-overrides-work) state that:
> Override blocks can also specify patterns to exclude from matches. If a file matches any of the excluded patterns, the configuration won't apply.

**Review by arcanis — COMMENTED** [review]
Forgot to press send 😅

File: packages/gatsby/gatsby-plugin-yarn-introspection/gatsby-node.js [hunk 3]
```diff
@@ -19,10 +19,12 @@ exports.sourceNodes = async ({actions, createNodeId, createContentDigest}, opts)
...
 
-  const data = await execute([
-    require.resolve(`@yarnpkg/monorepo/packages/yarnpkg-core/sources/Plugin.ts`),
+  const files = [
+    path.join(packageDirectory, `yarnpkg-core/sources/Plugin.ts`),
     ...indexList,
-  ]);
+  ];
+
+  const data = await execute(files);
```
[the remark this answers is not in the document]
Comment (paul-soporan) [reply packages/gatsby/gatsby-plugin-yarn-introspection/gatsby-node.js]: Leftovers from the previous implementation. I think the `path.join` call had something to do with the fact that `require.resolve` was returning virtuals but that was a long time ago and I don't remember the details.

**Pull request merged**